docs/design/node_taint_eviction.md

```diff
@@ -0,0 +1,78 @@
+# Tainted Node Eviction
+
+## Metadata
+
+* Authors: @johscheuer
+* Created: 2023-02-14
+* Updated: 2023-02-14
+
+## Background
+
+The operator is currently able to detect failures of Pods and replace them if required, but there is still a gap between an issue that manifests in issue that are detected by the operator and issues that could potentially happen.
+In Kubernetes it's common to use [taints](https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration) to indicate potential problems with a node or planned maintenance.
+This information could be consumed by the operator to proactively replace Pods that are running on node with a given set of taints.
+
+## General Design Goals
+
+The goal of this design is to implement a mechanism that allows the operator to replace Pods that are running on nodes with a taint.
+
+## Current Implementation
+
+We don't have a current implementation of this and the operator is currently not able to access any node information.
+The closest implementation to this is the [Replace Failed ProcessGroups reconciler](https://github.com/FoundationDB/fdb-kubernetes-operator/blob/main/controllers/replace_failed_process_groups.go), which could be extended for this.
+
+## Proposed Design
+
+We have to make sure that this feature can be enabled and disabled, e.g. if someone wants to run the operator inside a cluster without node access.
+The `AutomaticReplacementOptions` will be extended to support configuration of replacements for Pods that run on tainted nodes.
+The new field will be called `taints` and will contain a list, default would be empty, of taint configurations.
+That configuration will include the taint key, this has to be an exact match or a wildcard `*`, and the duration this taint must be present on the node.
+The `durationInSeconds` allow to specify how sensitive the operator should be to taints.
+If a wildcard is specified as key this will be used as default value, except there is an exact match defined for a taint key.
+
+```yaml
+...
+automaticReplacementOptions:
+  enabled: true
+  failureDetectionTimeSeconds: 7200
+  maxConcurrentReplacements: 1
+  enableNodeFailureDetection: true
+  taints:
+  - key: "example.org/maintenance"
+    durationInSeconds: 7200 # Ensure the taint is present for at least 2 hours before replacing Pods on a node with this taint
+  - key: "*" # The wildcard would allow to define a catch all configuration
+    durationInSeconds: 3600 # Ensure the taint is present for at least 1 hour before replacing Pods on a node with this taint
+```
+
+The actual implementation will extend the [ReplaceFailedProcessGroups](https://github.com/FoundationDB/fdb-kubernetes-operator/blob/main/internal/replacements/replace_failed_process_groups.go) to handle those cases.
```

Comment:
Rather than modifying this reconciler, could we add a process group condition indicating that a pod is on a tainted node, and make use of the existing logic for replacing process groups? That might involve less changes to the code, and allow more re-use.

Reply:
I was thinking about the same and the drawback of this approach would be that we are not able to replace all Pods on the tainted node "at once". I guess we can try that approach first before adding more complex logic.

Reply:
Updated the design to reflect, that we use the same replacement mechanism and just add a new condition for Pods running on a tainted node.